### ISSUE TITLE
[Bug] Adam 優化器使用 per-parameter gradient clipping 而非 global norm clipping

### ISSUE THREAD

**Issue by s9213712** [s9213712/minicnn#32]
## 問題描述

`src/minicnn/optim/adam.py` 中的 `grad_clip` 對每個參數的梯度個別裁剪：

```python
if self.grad_clip > 0.0:
    norm = float(np.linalg.norm(grad))
    if norm > self.grad_clip:
        grad = grad * (self.grad_clip / norm)
```

這是 **per-parameter clipping**，而 PyTorch 的 `clip_grad_norm_()` 以及大多數 ML 框架標準是對**所有參數的梯度向量整體計算 global norm** 再裁剪。

## 差異說明

- **Per-parameter clipping**（目前行為）：每個參數獨立限制自己的梯度範數為 `grad_clip`
- **Global norm clipping**（標準）：計算所有參數梯度的合併 L2 範數，若超過 `grad_clip` 則按相同比例縮放

Per-parameter clipping 會改變梯度方向（各參數縮放比例不同），在訓練 RNN 等梯度爆炸敏感的模型時行為與預期不符。

## 影響

- **嚴重性**: Major
- 使用 `grad_clip` 的訓練配置與 PyTorch 基線比較時結果不可重現
- `configs/` 中有多個 YAML 設定使用了 `grad_clip`

## 建議修復

```python
def step(self):
    self.t += 1
    # Global norm clipping
    if self.grad_clip > 0.0:
        all_grads = [p.grad for p in self.params if p.grad is not None]
        if all_grads:
            global_norm = float(np.sqrt(sum(np.sum(g**2) for g in all_grads)))
            if global_norm > self.grad_clip:
                scale = self.grad_clip / global_norm
                for p in self.params:
                    if p.grad is not None:
                        p.grad = p.grad * scale
    
    for i, p in enumerate(self.params):
        # ... 後續更新 ...
```

## 相關位置

- `src/minicnn/optim/adam.py:37-40`
- `src/minicnn/optim/adamw.py` — 可能有相同問題
